core/modules/qana/QservRestrictorPlugin.cc

```diff
@@ -102,6 +110,190 @@ lookupSecIndex(query::QueryContext& context,
+    }
+
+    // Build the QsRestrictor
+    query::QsRestrictor::Ptr restrictor = std::make_shared<query::QsRestrictor>();
+    if (restrictorType==SECONDARY_INDEX_IN) {
+        restrictor->_name = "sIndex";
+    }
+    else if (restrictorType==SECONDARY_INDEX_BETWEEN) {
+        restrictor->_name = "sIndexBetween";
+    }
+    // sIndex and sIndexBetween have parameters as follows:
+    // db, table, column, val1, val2, ...
+
+    css::PartTableParams const partParam = context.css->getPartTableParams(cr->db, cr->table);
+    // Get the director column name
+    std::string dirCol = partParam.dirColName;
+    if (cr->column == dirCol) {
+        // cr may be a column in a child table, in which case we must figure
+        // out the corresponding column in the child's director to properly
+        // generate a secondary index constraint.
+        std::string dirDb = partParam.dirDb;
+        std::string dirTable = partParam.dirTable;
+        if (dirTable.empty()) {
+            dirTable = cr->table;
+            if (!dirDb.empty() && dirDb != cr->db) {
+                LOGF_ERROR("dirTable missing, but dirDb is set inconsistently for %1%.%2%"
+                           % cr->db % cr->table);
+                return query::QsRestrictor::Ptr();
+            }
+            dirDb = cr->db;
+        } else if (dirDb.empty()) {
+            dirDb = cr->db;
+        }
+        if (dirDb != cr->db || dirTable != cr->table) {
+            // Lookup the name of the director column in the director table
+            dirCol = context.css->getPartTableParams(dirDb, dirTable).dirColName;
+            if (dirCol.empty()) {
+                LOGF_ERROR("dirCol missing for %1%.%2%" % dirDb % dirTable);
+                return query::QsRestrictor::Ptr();
+            }
+        }
+        LOGF_DEBUG("Using dirDb %1%, dirTable %2%, dirCol %3% as sIndex for %4%.%5%.%6%" %
+                   dirDb % dirTable % dirCol % cr->db % cr->table % cr->column);
+        restrictor->_params.push_back(dirDb);
+        restrictor->_params.push_back(dirTable);
+        restrictor->_params.push_back(dirCol);
+    } else {
+        LOGF_DEBUG("Using %1%, %2%, %3% as sIndex." % cr->db % cr->table % cr->column);
+        restrictor->_params.push_back(cr->db);
+        restrictor->_params.push_back(cr->table);
+        restrictor->_params.push_back(cr->column);
+    }
+    std::transform(values.begin(), values.end(),
+                   std::back_inserter(restrictor->_params), extractLiteral());
+    return restrictor;
+}
+
+/**  Create QSRestrictors which will use secondary index
+ *
+ *   @param context:  Context used to analyze SQL query, allow to compute
+ *                    columns names and find if they are in secondary index.
+ *   @param andTerm:  Intermediatre representation of a subset of a SQL WHERE clause
+ *
+ *   @return:         Qserv restrictors list
+ */
+std::shared_ptr<query::QsRestrictor::PtrVector> getSecIndexRestrictors(query::QueryContext& context,
```

Comment:
I think this is abuse of the `shared_ptr`. You can return vector by value and it will likely be more efficient.